--- frontend/main.c	2009-03-01 13:14:05.000000000 -0500
+++ frontend/main.c	2009-03-01 13:15:59.000000000 -0500
@@ -1174,6 +1174,8 @@
 
 
     faad_fprintf(stderr, " *********** Ahead Software MPEG-4 AAC Decoder V%s ******************\n\n", FAAD2_VERSION);
+    faad_fprintf(stderr, "                  Patched for SqueezeCenter\n");
+    faad_fprintf(stderr, "          http://bugs.slimdevices.com/show_bug.cgi?id=10602\n\n");
     faad_fprintf(stderr, " Build: %s\n", __DATE__);
     faad_fprintf(stderr, " Copyright 2002-2004: Ahead Software AG\n");
     faad_fprintf(stderr, " http://www.audiocoding.com\n");

