--- EV-orig.xs	2010-03-25 11:30:46.000000000 -0400
+++ EV.xs	2010-03-25 11:30:50.000000000 -0400
@@ -30,7 +30,8 @@
 #include "libev/ev.c"
 
 #ifndef _WIN32
-# include <pthread.h>
+// pthread not supported in OSX 10.3 and we don't need it anyway
+//# include <pthread.h>
 #endif
 
 #define e_loop(w) INT2PTR (struct ev_loop *, SvIVX ((w)->loop))
@@ -168,7 +169,8 @@
     {
       rv = newRV_noinc (w->self);
       sv_bless (rv, stash);
-      SvREADONLY_on (w->self);
+      // this conflicts with weaken under perl 5.8.1, seems safe to comment it out
+      //SvREADONLY_on (w->self);
     }
 
   return rv;
@@ -489,7 +491,7 @@
     SvREADONLY_on (sv);
   }
 #ifndef _WIN32
-  pthread_atfork (0, 0, ev_default_fork);
+//  pthread_atfork (0, 0, ev_default_fork);
 #endif
 }
 
