--- Makefile.PL.orig	2008-04-21 10:12:25.000000000 -0400
+++ Makefile.PL	2009-10-03 03:56:21.000000000 -0400
@@ -225,6 +225,17 @@
 print F $DEFINES,"\n";
 close F;
 
+# Patch to build statically
+@LIBS = ( '-liconv', '-lz', '-lm' );
+push @LIBS, ( '-ljpeg_s -lpng12_s -lgd_s -lfreetype_s' );
+
+my $MYEXTLIB = join( ' ', (
+    '../build/lib/libjpeg.a',
+    '../build/lib/libpng12.a',
+    '../build/lib/libgd.a',
+    '../build/lib/libfreetype.a',
+) );
+
 WriteMakefile(
     'NAME'	=> 'GD',
     'VERSION_FROM'	=> 'GD.pm',
@@ -250,6 +263,7 @@
     'ABSTRACT'  => 'Interface to Gd Graphics Library',
      $CAPI ? ('CAPI'      => 'TRUE') : (),
     'DEFINE'	=> $DEFINES,
+    'MYEXTLIB' => $MYEXTLIB,
 );
 
 exit 0;
