diff -ur GD-2.41-orig/GD/Image.pm.PLS GD-2.41/GD/Image.pm.PLS
--- GD-2.41-orig/GD/Image.pm.PLS	2009-11-10 16:28:06.000000000 -0500
+++ GD-2.41/GD/Image.pm.PLS	2009-11-10 16:28:24.000000000 -0500
@@ -218,6 +218,15 @@
     $class->_newFromJpeg($fh,@_);
 }
 
+sub newFromJpegScaled {
+    croak("Usage: newFromJpeg(class,filehandle,target_width,target_height,[truecolor])") unless @_>=4;
+    my($class) = shift;
+    my($f)     = shift;
+    my $fh = $class->_make_filehandle($f);
+    binmode($fh);
+    $class->_newFromJpegScaled($fh,@_);
+}
+
 !NO!SUBS!
 }
 
diff -ur GD-2.41-orig/GD.xs GD-2.41/GD.xs
--- GD-2.41-orig/GD.xs	2009-11-10 16:28:06.000000000 -0500
+++ GD-2.41/GD.xs	2009-11-10 16:28:36.000000000 -0500
@@ -346,6 +346,7 @@
 #define GDIMAGECREATEFROMPNG(x)  gdImageCreateFromPng((FILE*)x)
 #define GDIMAGECREATEFROMXBM(x)  gdImageCreateFromXbm((FILE*)x)
 #define GDIMAGECREATEFROMJPEG(x) gdImageCreateFromJpeg((FILE*)x)
+#define GDIMAGECREATEFROMJPEGSCALED(x,a,b) gdImageCreateFromJpegScaled((FILE*)x,a,b)
 #define GDIMAGECREATEFROMGIF(x)  gdImageCreateFromGif((FILE*)x)
 #define GDIMAGECREATEFROMWBMP(x) gdImageCreateFromWBMP((FILE*)x)
 #define GDIMAGECREATEFROMGD(x)   gdImageCreateFromGd((FILE*)x)
@@ -357,6 +358,7 @@
 #define GDIMAGECREATEFROMPNG(x) gdImageCreateFromPng(PerlIO_findFILE(x))
 #define GDIMAGECREATEFROMXBM(x) gdImageCreateFromXbm(PerlIO_findFILE(x))
 #define GDIMAGECREATEFROMJPEG(x) gdImageCreateFromJpeg(PerlIO_findFILE(x))
+#define GDIMAGECREATEFROMJPEGSCALED(x,a,b) gdImageCreateFromJpegScaled(PerlIO_findFILE(x),a,b)
 #define GDIMAGECREATEFROMGIF(x)  gdImageCreateFromGif(PerlIO_findFILE(x))
 #define GDIMAGECREATEFROMWBMP(x) gdImageCreateFromWBMP(PerlIO_findFILE(x))
 #define GDIMAGECREATEFROMGD(x) gdImageCreateFromGd(PerlIO_findFILE(x))
@@ -366,6 +368,7 @@
 #define GDIMAGECREATEFROMPNG(x) gdImageCreateFromPng(x)
 #define GDIMAGECREATEFROMXBM(x) gdImageCreateFromXbm(x)
 #define GDIMAGECREATEFROMJPEG(x) gdImageCreateFromJpeg(x)
+#define GDIMAGECREATEFROMJPEGSCALED(x,a,b) gdImageCreateFromJpegScaled(x,a,b)
 #define GDIMAGECREATEFROMGIF(x) gdImageCreateFromGif(x)
 #define GDIMAGECREATEFROMWBMP(x) gdImageCreateFromWBMP(x)
 #define GDIMAGECREATEFROMGD(x) gdImageCreateFromGd(x)
@@ -692,7 +695,7 @@
 	CODE:
 	  data = SvPV(imageData,len);
           ctx = newDynamicCtx(data,len);
-          RETVAL = (GD__Image) gdImageCreateFromJpegCtx(ctx);
+          RETVAL = (GD__Image) gdImageCreateFromJpegCtx(ctx, 0, 0);
           (ctx->gd_free)(ctx);
           if (items > 2) truecolor = (int)SvIV(ST(2));
 	  gd_chkimagefmt(RETVAL, truecolor);
@@ -701,6 +704,40 @@
 
 #endif
 
+#ifdef HAVE_JPEG
+GD::Image
+gdnewFromJpegDataScaled(packname="GD::Image", imageData, ...)
+	char *	packname
+	SV *    imageData
+	PROTOTYPE: $$$$;$
+        PREINIT:
+	  gdIOCtx* ctx;
+          char*    data;
+          STRLEN   len;
+#ifdef START_MY_CXT
+	  dMY_CXT;
+          unsigned int target_width, target_height;
+          int     truecolor = MY_CXT.truecolor_default;
+#else
+          unsigned int target_width, target_height;
+          int     truecolor = truecolor_default;
+#endif
+	CODE:
+	  data = SvPV(imageData,len);
+          ctx = newDynamicCtx(data,len);
+          
+          target_width = (unsigned int)SvIV(ST(2));
+          target_height = (unsigned int)SvIV(ST(3));
+          
+          RETVAL = (GD__Image) gdImageCreateFromJpegCtx(ctx, target_width, target_height);
+          (ctx->gd_free)(ctx);
+          if (items > 4) truecolor = (int)SvIV(ST(4));
+	  gd_chkimagefmt(RETVAL, truecolor);
+	OUTPUT:
+	  RETVAL
+
+#endif
+
 GD::Image
 gdnewFromWBMPData(packname="GD::Image", imageData, ...)
 	char *	packname
@@ -778,6 +815,35 @@
 
 #endif
 
+#ifdef HAVE_JPEG
+GD::Image
+gd_newFromJpegScaled(packname="GD::Image", filehandle, ...)
+	char *	packname
+	InputStream	filehandle
+	PROTOTYPE: $$$$;$
+        PREINIT:
+#ifdef START_MY_CXT
+	  dMY_CXT;
+          unsigned int target_width, target_height;
+          int     truecolor = MY_CXT.truecolor_default;
+#else
+          unsigned int target_width, target_height;
+          int     truecolor = truecolor_default;
+#endif
+	CODE:
+  	target_width = (unsigned int)SvIV(ST(2));
+    target_height = (unsigned int)SvIV(ST(3));
+    
+	  if (items > 4) truecolor = (int)SvIV(ST(4));
+    
+	  RETVAL = GDIMAGECREATEFROMJPEGSCALED(filehandle, target_width, target_height);
+	  
+	  gd_chkimagefmt(RETVAL, truecolor);
+	OUTPUT:
+          RETVAL
+
+#endif
+
 GD::Image
 gd_newFromWBMP(packname="GD::Image", filehandle)
 	char *	packname
